File: src/main/java/com/extendedae_plus/network/SPacketLabelList.java
package com.extendedae_plus.network;

import com.extendedae_plus.common.registry.menu.labelLink.MenuLabelLink;
import com.extendedae_plus.common.wireless.linkApi.RegistryLink;
import com.extendedae_plus.integration.IntegrationFTBTeams;
import com.extendedae_plus.network.base.EAEPNetworkPacket;
import com.extendedae_plus.network.base.SPacketGeneric;
import com.extendedae_plus.network.helper.HelperHandlerClient;
import net.minecraft.client.player.LocalPlayer;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.server.level.ServerPlayer;
import net.neoforged.neoforge.network.PacketDistributor;

import java.util.List;
import java.util.Objects;

@EAEPNetworkPacket("label_list")
public record SPacketLabelList(List<MenuLabelLink.LabelMapped> labels) implements SPacketGeneric {
    public static final StreamCodec<RegistryFriendlyByteBuf, SPacketLabelList> STREAM_CODEC = StreamCodec.composite(
            MenuLabelLink.LabelMapped.STREAM_CODEC.apply(ByteBufCodecs.list()), SPacketLabelList::labels,
            SPacketLabelList::new
    );

    public static void send(MenuLabelLink menu) {
        var placer = IntegrationFTBTeams.instance
                .getTeamUUID(menu.getPlayer().getUUID())
                .orElse(menu.getPlayer().getUUID());

        var serial = new int[]{Integer.MIN_VALUE};
        var labels = RegistryLink.getLabels().stream()
                .filter(label -> Objects.equals(label.data.placer(), placer) || label.data.placer() == null)
                .map(label -> new MenuLabelLink.LabelMapped(serial[0]++, label.data))
                .toList();
        menu.setLabels(labels);
        PacketDistributor.sendToPlayer((ServerPlayer) menu.getPlayer(), new SPacketLabelList(labels));
    }

    @Override
    public void handleClient(LocalPlayer player) {
        HelperHandlerClient.instance.labelList(this, player);
    }
}

File: src/main/kotlin/com/extendedae_plus/network/SPacketLabelList.kt
package com.extendedae_plus.network

import com.extendedae_plus.common.registry.menu.labelLink.MenuLabelLink
import com.extendedae_plus.common.registry.menu.labelLink.MenuLabelLink.LabelMapped
import com.extendedae_plus.common.wireless.linkApi.RegistryLink
import com.extendedae_plus.integration.helper.ManagerIntegration
import com.extendedae_plus.integration.impl.point.IntegrationFTBTeams
import com.extendedae_plus.network.helper.HandlersClient
import com.fish.fishlib.network.FishNetworkPacket
import com.fish.fishlib.network.PacketStreamCodec
import com.fish.fishlib.network.base.SPacketGeneric
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.server.level.ServerPlayer
import net.neoforged.neoforge.network.PacketDistributor

@FishNetworkPacket("label_list")
@JvmRecord
data class SPacketLabelList(val labels: MutableList<LabelMapped>) : SPacketGeneric {
    override val handlerClient get() = HandlersClient.LabelList()

    companion object {
        @PacketStreamCodec
        val streamCodec: StreamCodec<RegistryFriendlyByteBuf, SPacketLabelList> =
            StreamCodec.composite(
                LabelMapped.streamCodec.apply(ByteBufCodecs.list()),
                SPacketLabelList::labels,
                ::SPacketLabelList
            )

        @JvmStatic
        fun send(menu: MenuLabelLink) {
            val uuidPlayer = menu.player.uuid
            val uuidWrapped = ManagerIntegration<IntegrationFTBTeams>()
                ?.getTeamUUID(uuidPlayer)
                ?: uuidPlayer

            var serial = Int.MIN_VALUE
            val labels = RegistryLink.labels
                .filter { it.data.placer == uuidWrapped || it.data.placer == null }
                .mapTo(ArrayList()) { LabelMapped(serial++, it.data) }
            menu.setLabels(labels)
            PacketDistributor.sendToPlayer(menu.player as ServerPlayer, SPacketLabelList(labels))
        }
    }
}